composer/trainer/trainer.py

```diff
@@ -140,6 +246,140 @@ def _unpack_evaluators(
+            which may cause uncaught Cuda Out of Memory errors. In this case, please manually
+            set grad_accum explicitly to an integer instead.
+            """),
+                      category=UserWarning)
+    # Cannot use adaptive grad accum on CPU
+    if isinstance(device, DeviceCPU) and adaptive_grad_accum:
+        raise ValueError("Cannot use adaptive grad_accum on CPU. Please set grad_accum >= 1")
+    # grad_accum should be int as we've already handled "auto" case
+    if isinstance(grad_accum, str):
+        raise ValueError("grad_accum must be an int or ``auto``")
+    return grad_accum, adaptive_grad_accum
+
+
+def _is_cuda_oom(e: RuntimeError):
+    """Determines if error is CUDA Out of Memory and if adaptive_grad_accum is enabled."""
+    return "CUDA out of memory" in str(e)
+
+
+def _unpack_device(device: Optional[Union[str, Device]], deepspeed_enabled: bool):
+    if not device:
+        device = DeviceGPU() if torch.cuda.is_available() else DeviceCPU()
+    elif isinstance(device, str):
+        if device.lower() == 'cpu':
+            device = DeviceCPU()
+        elif device.lower() == 'gpu':
+            device = DeviceGPU()
+        else:
+            raise ValueError(f'device ({device}) must be one of (cpu, gpu).')
+    return device
+
+
+def _unpack_seed(seed: Optional[int], device: Device):
+    if not seed:
+        seed = reproducibility.get_random_seed()
+
+    # Ensure that each process has a seed = rank_zero_seed + global_rank
+    # This "deterministically different" seed behavior is required to be able
+    # to restore seeds when resuming form checkpoints, since only the
+    # `rank_zero_seed` is stored on state.
+    if seed < 0 or seed > reproducibility.MAX_SEED:
+        raise ValueError(f"Invalid seed: {seed}. It must be on [0; 2**32 - 1)")
+
+    # using int64 to prevent overflow
+    rank_zero_seed = device.tensor_to_device(torch.tensor([seed], dtype=torch.int64))
+    dist.broadcast(rank_zero_seed, src=0)
+    rank_zero_seed = rank_zero_seed.item()
+    assert isinstance(rank_zero_seed, int)
+    seed = rank_zero_seed + dist.get_global_rank()
+    log.info(f"Setting seed to {seed}")
+    return rank_zero_seed, seed
+
+
+def _unpack_max_duration(max_duration: Union[int, str, Time]):
+    if isinstance(max_duration, str):
+        max_duration = Time.from_timestring(max_duration)
+    elif isinstance(max_duration, int):
+        max_duration = Time.from_epoch(max_duration)
+    return max_duration
+
+
+def _unpack_ddp_sync_strategy(ddp_sync_strategy: Optional[Union[str, DDPSyncStrategy]], find_unused_parameters: bool):
+    if ddp_sync_strategy is None:
+        if find_unused_parameters:
+            ddp_sync_strategy = DDPSyncStrategy.FORCED_SYNC
+        else:
+            ddp_sync_strategy = DDPSyncStrategy.SINGLE_AUTO_SYNC
+    else:
+        ddp_sync_strategy = DDPSyncStrategy(ddp_sync_strategy)
+    return ddp_sync_strategy
+
+
+def _unpack_profiler(
```

Comment:
This maybe outside the scope, but i would be in favor of setting a:
```
trainer = Trainer(
    enable_profiler: bool = True
)
```

and then any other fancy configuration someone should just create and pass in the object (e.g. unflatten)?

Reply:
Yeah for example we could have a `ProfilerConfig` class that has all these other params (I know this is basically hparams but we do want to explicitly avoid hparams use here)

Reply:
Agree it would be nice to unflatten it. Tracking [here](https://mosaicml.atlassian.net/browse/CO-296). There are a few ways to address this -- one is to pass in the profiler object directly (would require a bit of re-writing, as the `Profiler` constructor currently requires the state, or as @ajaysaini725 mentioned, pass in a config/hparams class.